src/sentry/models/dashboard_widget.py

```diff
@@ -114,6 +118,61 @@ class Meta:
     __repr__ = sane_repr("widget", "type", "name")
 
 
+class DashboardWidgetQueryOnDemandManager(Manager):
+    def filter_enabled(self):
+        return super().get_queryset().filter(extraction_reason__startswith=_ON_DEMAND_ENABLED_KEY)
+
+
+@region_silo_only_model
+class DashboardWidgetQueryOnDemand(Model):
+    """
+    Tracks on_demand state and values for dashboard widget queries.
+    Only a subset of dashboard widget queries have conditions or columns that would
+    require on-demand extraction, and others are simply not applicable (eg. different dataset).
+    """
+
+    __relocation_scope__ = RelocationScope.Organization
+
+    dashboard_widget_query = FlexibleForeignKey("sentry.DashboardWidgetQuery")
+
+    spec_hashes = ArrayField()
+
+    class OnDemandExtractionState(models.TextChoices):
+        DISABLED_NOT_APPLICABLE = "disabled:not_applicable", gettext_lazy("disabled:not_applicable")
+        """ This widget does not have on-demand metrics needing extraction. """
+        DISABLED_PREROLLOUT = "disabled:pre_rollout", gettext_lazy("disabled:pre_rollout")
+        """ This represents a pre-filled on-demand value to do load estimates before enabling extraction. """
+        DISABLED_MANUAL = "disabled:manual", gettext_lazy("disabled:manual")
+        """ The widget was manually disabled by a user """
+        DISABLED_SPEC_LIMIT = "disabled:spec-limit", gettext_lazy("disabled:spec-limit")
+        """ This widget query was disabled during rollout due to the organization reaching it's spec limit. """
```

Comment:
Whenever we have some documentation we should explain what this means to the customer and what their options are.

Reply:
It'll be in the technical doc, it's mostly [written down here at the moment](https://www.notion.so/sentry/Adding-on-demand-state-to-dashboard-widgets-5f633701b503498bb84b4631fe51e352) but I changed my mind on a couple things.